### DISCUSSION TITLE
[Question]: PTZ and mask

### DISCUSSION THREAD

**Discussion by ramirez22-fr** [blakeblackshear/frigate#13286] [Ask A Question]
### What is your question:

Hello.
First, many thanks for developer and all contributor : it's really an amazing job !

Devices:
Computer i5 with Proxmox, only one VM with home assistant and frigate HACS/Integration/add-on.
Coral USB
2 PTZ cam (ASECAM AJ8MPTZ5X : possible to add to the PTZ compliance list :) )

All soft are in the latest version

All work perfectly:
- Stream
- Reccord
- Human recongnizion
- PTZ (move and preset)
- Notification (with blueprint "Frigate notifications")

The camera are in garden, so during windly days, the Coral is at its limit.
To reduce Coral sollicitation, I set motion mask.

But of course, masks are not the same for each preset ...
I disabled motion sensor during move of the camera (few seconds), to limit TPU acces.

But is exist a way to activate a mask and desactivate another one on a camera, to match with the scene/preset ?

**Comment by hawkeye217** [comment]
No, there is currently no way to do this. Zones and masks in frigate presume a static image.

**Comment by ramirez22-fr** [comment]
Ah ... too bad.

I don't know if its easy to implement this, and I don't saw a request about this.
Do you think it's a good idea ?

But I'm not developper, maybe it's really hard ...

Wait and see

**Reply by hawkeye217** [reply]
The ONVIF protocol does not have a way of reporting if a camera's current position is a preset or not. So yes, this would be hard to accomplish.

**Comment by ramirez22-fr** [comment]
Not necessary automatic switch. 
I imagine something like this (automation under home assistant for example):
- when change position has to be done, disable motion detection 
- call the new PTZ preset
- Wait a few moment (timer manually adjust with test for each cam) for positionning
- disable mask of the old preset, enable mask of the new position and enable again motion detection

But it's easy to say when no coding abilities 🤣

**Reply by abate** [reply]
I face the same issue. The solution above is not that bad. basically for each camera, we should be able to set a mask with an id associated to each preset. Instead of an array, the structure should be an array of arrays. and in the UI one more drop list when creating a mask. I agree. easier said than done.

**Reply by hawkeye217** [reply]
As I said above, there's no way for Frigate to know what "preset" your camera is set to as this information is not part of the ONVIF standard. Without this data, support would be a nightmare as there's no way to standardize this feature so that it works for everyone's camera.

**Reply by abate** [reply]
I see. I just read the onvif spec regarding this. You can create, remove, goto, etc. But there is no way to ask which is the current preset.  So if frigate was to apply a mask it would not able to know if the current mask is the correct one for the current position.

Moreover, since cameras with auto tracking, auto focus, whatever and move independently from frigate, and the only way to keep the state of the camera is to check the coordinates, it would be indeed complicated to guess when to apply the correct mask.

This is a bummer. Thank you for your answer !

**Comment by Jpereyra316** [comment]
Why do need to "ask" which preset the camera is on? I wouldn't care which preset it's on. It should just follow the preset it's told to go to.

It would be nice if HA could control a camera by either specifying PTZ parameters or a preset. If HA commands to set to preset1, HA could also command to switch to mask1 (or whatever user calls it). Basically, control of preset (PTZ) and mask would be handled by HA. This would be useful if an event such as an open/close door/window occurs, then HA would command the camera to set the preset and mask that would record that event's location. Then HA can switch back to some other default preset and mask.

Another instance example is:
I have a bed occupancy detector I built for my baby daughter. I would like to switch the camera on/off based on whether she's in the crib or not. HA knows occupancy, but I don't see how to control the PTZ for the camera.

I wouldn't mind helping develop this.

Thoughts?

**Reply by hawkeye217** [reply]
That sounds great in principle, but when Frigate or HA starts/restarts, how do you know which mask to apply if you don't know what camera preset is currently set? Also, how would you determine if the camera has actually moved to the correct preset and is done moving? 

Without the camera's ability to report preset data through ONVIF, there is much room for error. Without standardization, this makes support a nightmare. So it's unlikely we would ever consider any kind of implementation of this in Frigate.

**Reply by Jpereyra316** [reply]
That's fair. I see your point. It's difficult to verify if the request completed successfully. 

It would be painful to work around such limitation. 

I just learned about Frigate yesterday and got it working today with a couple cams so I'm excited to continue exploring.

**Reply by Jpereyra316** [reply]
In case someone with a similar situation ends up here...

I have Tapo cameras that support PTZ. I was able to use the tapo control integration (https://github.com/JurajNyiri/HomeAssistant-Tapo-Control) to control pan, tilt and preset from HA to achieve what I needed. However, I am NOT using a mask that depends on the preset set at the moment. 

My automation is implemented in Node-Red add-on. When my condition is true, preset 1 is triggered; if false, preset 2. I was also able to catch failures to send the preset change request in Node-red to retry (this is my "verification", though it doesn't guarantee the position is exact, it's close enough). Since the current state of PTZ is irrelevant in this situation, it's working well, and I don't see an issue if camera, HA or Frigate restart.

**Comment by nurtext** [comment]
@ramirez22-fr Regarding the ASECAM AJ8MPTZ5X: Is it working PTZ only or is autotracking supported?

**Comment by uncle-buzz** [comment]
I second this demand, it would be great to allow HA or any third party to send a command via MQTT or another way to activate/deactivate some masks (motion and others). Obviously, there will be some problematic scenarios to deal with (reset, startup, not responding cam...) but it will be the job of the third party and not of Frigate to deal with that, just let us activate/deactivate masks on demand and don't support what's going on after.
Having some masks associated with preset PTZ positions (which are static images if everything is working as expected) would be fantastic, we only need a way to activate/deactivate masks on the fly, association with preset will be our stuff.
Actually, it would be good to set not only the mask but also any detection parameters since the zoom will change the size in pixels of the subject, an API to dynamically change those parameters outside of Frigate without restarting it would be really good.

**Reply by hawkeye217** [reply]
As I said above, there's no way for Frigate to know what "preset" your camera is set to as this information is not part of the ONVIF standard. Without this data, support would be a nightmare as there's no way to standardize this feature so that it works for everyone's camera.

**Reply by uncle-buzz** [reply]
Frigate doesn't need to know anything if this is managed via API from a third party, the third party has to deal with that so you don't have to support anything, just allow access to enable/disable some masks dynamically as some other parameters on the fly.
If I ask my camera to go to a preset, and I know the set of parameters for this static position, I would like to activate those set of parameters.
I don't understand why you want to control everything in Frigate, the masks and detection parameters are written manually in the config file, Frigate uses them without any control over them, so there will be no difference to my mind. The reason why the parameters are modified is not the matter of Frigate.

**Reply by hawkeye217** [reply]
That sounds great in principle, but when Frigate or HA starts/restarts, how do you know which mask to apply if you don't know what camera preset is currently set? Also, how would you determine if the camera has actually moved to the correct preset and is done moving?

Without the camera's ability to report preset data through ONVIF, there is much room for error. Without standardization, this makes support a nightmare. So it's unlikely we would ever consider any kind of implementation of this in Frigate.

**Reply by uncle-buzz** [reply]
What if I set a mask and a zone on a PTZ camera, and I manually change the zoom or orientation after the start of Frigate? Is there any check of Frigate when I manually command the PTZ camera even through the ONVIF support of Frigate? Of course the masks and zones won't be anymore aligned and wrong things will happens.

Why do you want to add more control if I "manually" change some parameters from an API?

Allowing to change parameters via an API doesn't mean you have to control any modification.

If I edit the config file with a script and restart Frigate, I have the same result, appart that I have to restart frigate, the latence of any change is big and I lost all other cameras while the restart. 

I feel you don't read and copy/paste the same argument even if I have already pointed out that it's not a problem.

**Reply by hawkeye217** [reply]
We have already considered everything you have written. What you are asking for is slightly different from the OP, and would mean a fundamental redesign of the way Frigate uses masks and zones in object tracking and historical review. Frigate was designed with a static camera frame in mind. There are no plans to change this. 

Additionally, supporting users with such kind of variable configuration with a moving camera frame would be nearly impossible. 

I am not trying to be combative, I just want to make clear that we have considered these things at length. Thanks for your interest and enthusiasm, though.

**Reply by NickM-27** [reply]
To be clear there is a pinned feature request for dynamic zones and masks https://github.com/blakeblackshear/frigate/issues/3621

The association with camera presets likely won't be implemented for the reasons described above.

**Reply by hawkeye217** [reply]
Important to note that the dynamic zones and masks feature request would be to enable/disable certain masks and zones, not redefine the coordinates of existing masks and zones.

**Reply by Jpereyra316** [reply]
Will the enable/disable of the different masks and zones be exposed to HA via API/MQTT or will it just be based on time?

It would be really nice if it was exposed. That would solve this issue for those that want to use it this way.
If it's available to HA, I would define all possible zones/masks for all presets. Then, whenever I change the preset from HA, I can also enable/disable the zones/masks for that preset.

**Reply by NickM-27** [reply]
it would likely be dynamic via mqtt like most other things in Frigate

**Reply by hawkeye217** [reply]
If implemented, yes, it would likely expose the state/set via MQTT.

Though again to be clear, it would only enable/disable zones, not change the coordinates of existing zones. Each of your "dynamic" zones would need to be named differently.

**Reply by uncle-buzz** [reply]
Using an API like "frigate/<camera_name>/motion_mask/<mask_name>/set" to enable/disable a motion mask (and the same for zones) would meet the needs perfectly.
Enabling/disabling depending on time, depending on a preset PTZ or any other reason would be the responsibility of the caller of the API, not Frigate, so from Frigate's point of view there is no link to the PTZ preset which is only one usecase from many.
Of course, masks and zones should have been created before and won't be calculated by Frigate (which is a specific case with PTZ cameras, but not in other cases like time).

**Reply by uncle-buzz** [reply]
> I am not trying to be combative, I just want to make clear that we have considered these things at length. Thanks for your interest and enthusiasm, though.

As a non-native English speaker, sorry if my words seem aggressive, I really appreciate Frigate as it is and all the work that has been done!
